src/ApacheOrcDotNet.OptimizedReader/Buffers/BaseColumnBuffer.cs

```diff
@@ -0,0 +1,285 @@
+using System.Threading.Tasks;
+
+namespace ApacheOrcDotNet.OptimizedReader.Buffers
+{
+    [SkipLocalsInit]
+    public abstract class BaseColumnBuffer<TOutput>
+    {
+        private readonly long[] _numericStreamBuffer;
+        private readonly byte[] _byteStreamBuffer;
+        private readonly byte[] _boolStreamBuffer;
+
+        private protected readonly IByteRangeProvider _byteRangeProvider;
+        private protected readonly OrcContext _context;
+        private protected readonly OrcColumn _column;
+        private protected readonly TOutput[] _values;
+
+        private protected readonly ArrayPool<byte> _pool;
+
+        private protected int _numValuesRead;
+
+        private StreamRange _lastRange;
+        private int _lastRangeLength;
+
+        public BaseColumnBuffer(IByteRangeProvider byteRangeProvider, OrcContext context, OrcColumn column)
+        {
+            _byteRangeProvider = byteRangeProvider;
+            _context = context;
+            _column = column;
+            _values = new TOutput[_context.MaxValuesToRead];
+
+            _pool = ArrayPool<byte>.Create(15 * 1024 * 1024, 8);
+
+            _numericStreamBuffer = new long[1000];
+            _byteStreamBuffer = new byte[1000];
+            _boolStreamBuffer = new byte[1000];
+        }
+
+        public OrcColumn Column => _column;
+        public ReadOnlySpan<TOutput> Values => _values.AsSpan().Slice(0, _numValuesRead);
+
+        public abstract Task LoadDataAsync(int stripeId, ColumnDataStreams streams);
+        public abstract void Fill();
+
+        public void Reset() => _numValuesRead = 0;
+
+        private protected StreamDetail GetStripeStream(IEnumerable<StreamDetail> columnStreams, StreamKind streamKind, bool isRequired = true)
+        {
+            var stream = columnStreams.SingleOrDefault(stream =>
+                stream.StreamKind == streamKind
+            );
+
+            if (isRequired && stream == null)
+                throw new InvalidDataException($"The '{streamKind}' stream must be available");
+
+            return stream;
+        }
+
+        private protected void ReadByteStream(StreamDetail stream, ReadOnlySpan<byte> buffer, int length, Span<byte> outputValues, out int numValuesRead)
+        {
+            numValuesRead = 0;
+
+            if (stream == null)
+                return;
+
+            var numSkipped = 0;
+            var bufferReader = new BufferReader(ResizeBuffer(stream, buffer, length));
+
+            while (!bufferReader.Complete)
+            {
+                var numByteValuesRead = OptimizedByteRLE.ReadValues(ref bufferReader, _byteStreamBuffer);
+
+                for (int idx = 0; idx < numByteValuesRead; idx++)
+                {
+                    if (numSkipped++ < stream.Positions.ValuesToSkip)
+                        continue;
+
+                    outputValues[numValuesRead++] = _byteStreamBuffer[idx];
+
+                    if (numValuesRead >= outputValues.Length)
```

Comment:
Wouldn't this make things a bit harder to read in here?

I ask because we had requests in previous code reviews that we did the opposite (e.g.: moving multiple checks from the `while` declaration into the loop block).

Reply:
If we know the preconditions for a loop up front I think it's reasonable to include those in the `for` loop itself.  I'll often pull complex `IsComplete` decision making out to an `if/break`, but when we're thinking about ranges of buffers I think it's better for that to go in the loop declaration.